src/npctalk.cpp

```diff
@@ -1971,24 +1971,20 @@ void talk_effect_fun_t::set_u_sell_item( const std::string &item_name, int cost,
         npc &p = *d.beta;
         player &u = *d.alpha;
         item old_item = item( item_name, calendar::turn );
-        if( u.has_charges( item_name, count ) ) {
+        if( old_item.count_by_charges() && u.has_charges( item_name, count ) ) {
             u.use_charges( item_name, count );
+            old_item.mod_charges( count - 1 );
```

Comment:
This makes no sense. The charges of `old_item` are pretty much unknown (the default charges are used when it's created, which can be *anything*). `mod_charges` increases that number by `count - 1` (or decreases if its a negative value). But usually it will do nothing as `count` will be 1 and so it's called as `mod_charges(0)`.

Anyway, try this:
```C++
for( const item &it : u.use_charges( item_name, count ) ) {
    p.i_add( it );
}
```
Similar for `use_amount`.

And absolutely get rid of `old_item`. It is just a brand new item, it did not exist before the exchange, and the exchange should not create new items, so it's not needed.

Reply:
Without `old_item`, how shall I determine whether the item being traded is `count_by_charges()` or not? There is no item object passed into the function, only an item name.

Reply:
> Without old_item, how shall I determine whether the item being traded is count_by_charges() or not?

Use `item::count_by_charges` :-)

There is another overload of that function that is static and takes the item id as parameter. E.g. `item::count_by_charges( item_name )`.

Reply:
Aha, sweet! Thanks, I will try that.